moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,548 @@
+   * matching calls to [endArray] or [endObject]. Use this to compose JSON adapters
+   * without nesting.
+   *
+   *
+   * For example, the following creates JSON with nested arrays: `[1,[2,3,4],5]`.
+   *
+   * ```java
+   * JsonAdapter<List<Integer>> integersAdapter = ...
+   * public void writeNumbers(JsonWriter writer) {
+   *   writer.beginArray();
+   *   writer.value(1);
+   *   integersAdapter.toJson(writer, Arrays.asList(2, 3, 4));
+   *   writer.value(5);
+   *   writer.endArray();
+   * }
+   * ```
+   *
+   *
+   * With flattening we can create JSON with a single array `[1,2,3,4,5]`:
+   *
+   * ```java
+   * JsonAdapter<List<Integer>> integersAdapter = ...
+   *
+   * public void writeNumbers(JsonWriter writer) {
+   *   writer.beginArray();
+   *   int token = writer.beginFlatten();
+   *   writer.value(1);
+   *   integersAdapter.toJson(writer, Arrays.asList(2, 3, 4));
+   *   writer.value(5);
+   *   writer.endFlatten(token);
+   *   writer.endArray();
+   * }
+   * ```
+   *
+   *
+   * This method flattens arrays within arrays:
+   *
+   * Emit:       `[1, [2, 3, 4], 5]`
+   * To produce: `[1, 2, 3, 4, 5]`
+   *
+   * It also flattens objects within objects. Do not call [name] before writing a flattened
+   * object.
+   *
+   * Emit:       `{"a": 1, {"b": 2}, "c": 3}`
+   * To Produce: `{"a": 1, "b": 2, "c": 3}`
+   *
+   *
+   * Other combinations are permitted but do not perform flattening. For example, objects inside of
+   * arrays are not flattened:
+   *
+   * Emit:      ` [1, {"b": 2}, 3, [4, 5], 6]`
+   * To Produce: `[1, {"b": 2}, 3, 4, 5, 6]`
+   *
+   *
+   * This method returns an opaque token. Callers must match all calls to this method with a call
+   * to [endFlatten] with the matching token.
+   */
+  @CheckReturnValue
+  public fun beginFlatten(): Int {
+    val context = peekScope()
+    check(!(context != JsonScope.NONEMPTY_OBJECT && context != JsonScope.EMPTY_OBJECT && context != JsonScope.NONEMPTY_ARRAY && context != JsonScope.EMPTY_ARRAY)) { "Nesting problem." }
+    val token = flattenStackSize
+    flattenStackSize = stackSize
+    return token
+  }
+
+  /** Ends nested call flattening created by [beginFlatten].  */
+  public fun endFlatten(token: Int) {
+    flattenStackSize = token
+  }
+
+  /**
+   * Returns a [JsonPath](http://goessner.net/articles/JsonPath/) to the current location
+   * in the JSON value.
+   */
+  @get:CheckReturnValue
+  public val path: String
+    get() = JsonScope.getPath(stackSize, scopes, pathNames, pathIndices)
+
+  /** Returns the tag value for the given class key.  */
+  @CheckReturnValue
+  public fun <T> tag(clazz: Class<T>): T? {
+    @Suppress("UNCHECKED_CAST")
+    return if (tags == null) null else tags!![clazz] as T?
+  }
+
+  /** Assigns the tag value using the given class key and value.  */
+  public fun <T> setTag(clazz: Class<T>, value: T) {
+    require(clazz.isAssignableFrom(value!!::class.java)) { "Tag value must be of type " + clazz.name }
```

Comment:
Why the `!!` here? Also please use a string template

Reply:
Expression in a class literal has a nullable type 'T', use !! to make the type non-nullable

Reply:
There's another appropriate way to instance check this from the `clazz` and I think it would be a good exercise for yourself to figure out for yourself how to do so.

Reply:
Like this? Make T extends Any?
```kotlin
public fun <T : Any> setTag(clazz: Class<T>, value: T)
```

Reply:
No

Reply:
7d633c1d98bd7dcf120f2fdd08856bd9ed6bbd0c

Reply:
Maybe can use this? 

```kotlin
value::class.isSubclassOf(clazz.kotlin)
```